torch_mutual_information/rnnt.py

```diff
@@ -125,3 +126,304 @@ def rnnt_loss_simple(lm: Tensor,
+    py -= normalizers
+    px = px.contiguous()
+    py = py.contiguous()
+
+    return (px, py)
+
+
+def rnnt_loss(joint: Tensor,
+             symbols: Tensor,
+             termination_symbol: int,
+             boundary: Tensor = None) -> Tensor:
+    """
+    A normal RNN-T loss, which uses a 'joiner' network output as input, i.e. a 4 dimensions tensor.
+
+    Args:
+      joint: The output of joiner network, with shape
+             (B, T, S + 1, C), i.e. batch, time_seq_len, symbol_seq_len+1, num_classes
+     symbols: The symbol sequences, a LongTensor of shape [B][S], and elements in {0..C-1}.
+              termination_symbol: the termination symbol, with 0 <= termination_symbol < C
+     termination_symbol: The termination symbol, with 0 <= termination_symbol < C
+     boundary: a LongTensor of shape [B, 4] with elements interpreted as
+        [begin_symbol, begin_frame, end_symbol, end_frame] that is treated as [0, 0, S, T]
+        if boundary is not supplied.
+        Most likely you will want begin_symbol and begin_frame to be zero.
+
+    Returns:
+      A Tensor of shape (B,), containing the total RNN-T loss values for each element
+      of the batch (like log-probs of sequences).
+    """
+    px, py = get_rnnt_logprobs_joint(joint, symbols, termination_symbol)
+    return mutual_information_recursion(px, py, boundary)
+
+
+def monotonic_lower_bound(src, s_range, max_symbol):
+    dest = torch.zeros_like(src)
+    for i in range(src.shape[0]):
+        min_value = max_symbol + 1
+        dest[i][src.shape[1] - 1] = max_symbol - s_range
+        for j in range(src.shape[1] - 2, -1, -1):
+            min_value = max(min(src[i][j], min_value), dest[i][j+1] - s_range + 1)
+            dest[i][j] = min_value
+        dest[i][0] = 0
+        for j in range(1, dest.shape[1]):
+            dest[i][j] = min(dest[i][j-1] + s_range - 1, dest[i][j])
+        assert dest[i][src.shape[1] - 1] == max_symbol - s_range 
+    return dest
+
+
+def get_pruning_ranges(px_grad: torch.Tensor, py_grad: torch.Tensor, s_range: int):
+    """
+    Get the pruning ranges for normal rnnt loss according to the grad of rnnt_loss_simple.
+
+    Args:
+      px_grad: The gradient of px, see docs in mutual_information_recursion for more details of px.
+      py_grad: The gradient of py, see docs in mutual_information_recursion for more details of py.
+      s_range: How many symbols to keep for each frame.
+
+    Returns:
+      A tensor contains the kept symbols id for each frame, with shape (B, T, s_range).
+    """
+    (B, S, T1) = px_grad.shape
+    T = T1 - 1;
+    assert py_grad.shape == (B, S + 1, T)
+    assert s_range <= S
+
+    px_pad = torch.zeros((B, 1, T + 1), dtype=px_grad.dtype, device=px_grad.device)
+    py_pad = torch.zeros((B, S + 1, 1), dtype=py_grad.dtype, device=py_grad.device)
+    tot_grad = torch.cat((px_grad, px_pad), dim=1) + torch.cat((py_grad, py_pad), dim=2) # (B, S + 1, T + 1)
```

Comment:
There should be a cumulative sum somewhere in here.  I may have been unclear in my initial description.
BTW I think Torch only supports inclusive sum, not exclusive-sum.  You will probably have to concatenate zeros and then the output of inclusive-sum.